File: samples/scala-eventsourced-counter/src/main/scala/com/example/actions/CounterTopicSubscriptionAction.scala
package com.example.actions

import com.akkaserverless.scalasdk.action.Action
import com.akkaserverless.scalasdk.action.ActionCreationContext
import com.google.protobuf.empty.Empty
import org.slf4j.LoggerFactory

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** An action. */
class CounterTopicSubscriptionAction(creationContext: ActionCreationContext)
    extends AbstractCounterTopicSubscriptionAction {
  private val logger = LoggerFactory.getLogger(getClass())
  /** Handler for "Increase". */
  override def increase(increased: Increased): Action.Effect[Empty] = {
    logger.info("Received increase event: " + increased.toString())
    effects.noReply
  }

  /** Handler for "Decrease". */
  override def decrease(decreased: Decreased): Action.Effect[Empty] = {
    logger.info("Received increase event: " + decreased.toString())
    effects.noReply
  }
}

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterTopicSubscriptionAction.java
package com.example.actions;

import com.example.CounterEvent;
import com.example.CounterEvent.ValueIncreased;
import com.example.CounterEvent.ValueMultiplied;
import com.example.Counter;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
// tag::class[]
import kalix.javasdk.action.Action;
import kalix.springsdk.annotations.Subscribe;

public class CounterTopicSubscriptionAction extends Action {

    private Logger logger = LoggerFactory.getLogger(CounterTopicSubscriptionAction.class);

    @Subscribe.Topic(value = "counter-events") // <1>
    public Action.Effect<Confirmed> onTopicEvent(CounterEvent event){ // <2>
        if (event instanceof ValueIncreased){
            logger.info("Received increased event: " + event.toString());
        } else if (event instanceof ValueMultiplied){
            logger.info("Received multiplied event: " + event.toString());
        }
        return effects().reply(Confirmed.defaultInstance()); // <3>
    }
}
// end::class[]